libtextsecure/message_receiver.js

```diff
@@ -984,45 +981,39 @@ MessageReceiver.prototype.extend({
 
     let conversation;
     try {
-      conversation = window.ConversationController.get(envelope.source);
+      conversation = await window.ConversationController.getOrCreateAndWait(envelope.source, 'private');
     } catch (e) {
       window.log.info('Error getting conversation: ', envelope.source);
     }
-
-    if (
-      envelope.preKeyBundleMessage &&
-      envelope.preKeyBundleMessage.type ===
-      textsecure.protobuf.PreKeyBundleMessage.Type.FRIEND_REQUEST
-    ) {
-      return this.handleFriendRequestMessage(envelope, content.dataMessage);
-    } else if (
-      envelope.type === textsecure.protobuf.Envelope.Type.CIPHERTEXT ||
-      // We also need to check for PREKEY_BUNDLE aswell if the session hasn't started.
-      // ref: libsignal-protocol.js:36120
-      envelope.type === textsecure.protobuf.Envelope.Type.PREKEY_BUNDLE
-    ) {
-      // If we get a cipher and we're already friends
-      // then we set our key exchange to complete
-      if (conversation && conversation.isFriend()) {
-        await conversation.setKeyExchangeCompleted();
-      }
+    if (envelope.type === textsecure.protobuf.Envelope.Type.FRIEND_REQUEST) {
+      conversation.onFriendRequestReceived();
     }
 
-    if (content.syncMessage) {
+    if (content.preKeyBundleMessage) {
+      const preKeyBundleMessage =
+        this.decodePreKeyBundleMessage(content.preKeyBundleMessage);
+      await this.savePreKeyBundleMessage(
+        envelope.source,
+        preKeyBundleMessage
+      );
+      return this.handleDataMessage(envelope, content.dataMessage, 'friend-request');
+    }
+    if (content.syncMessage)
       return this.handleSyncMessage(envelope, content.syncMessage);
-    } else if (content.dataMessage) {
+    if (content.dataMessage)
       return this.handleDataMessage(envelope, content.dataMessage);
-    } else if (content.nullMessage) {
+    if (content.nullMessage)
       return this.handleNullMessage(envelope, content.nullMessage);
-    } else if (content.callMessage) {
+    if (content.callMessage)
       return this.handleCallMessage(envelope, content.callMessage);
-    } else if (content.receiptMessage) {
+    if (content.receiptMessage)
       return this.handleReceiptMessage(envelope, content.receiptMessage);
-    }
-    this.removeFromCache(envelope);
-    if (envelope.preKeyBundleMessage) return null;
 
-    throw new Error('Unsupported content message');
+    this.removeFromCache(envelope);
+    // TODO: The empty friend request response could get lost
+    // Need to trigger this event whenever a message is received in the pending state
+    conversation.onFriendRequestAccepted();
```

Comment:
shouldn't this be in an else function for the envelope type check above?
```
if (envelope.type === textsecure.protobuf.Envelope.Type.FRIEND_REQUEST) {
  conversation.onFriendRequestReceived();
} else {
  conversation.onFriendRequestAccepted();
}
```

Reply:
Hm tru, I wonder if there's a more sensible place to put it if we are just going to trigger for every message then? Like onNewMessage or something?

Reply:
I think it's fine inside `innerHandleContent`.